File: code/design-patterns/behavioral/template-method/go/template_method.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// OTP is the abstract base class
type OTP interface {
	Process()
	generateOTP(length int) string
	saveOTP(otp string)
	prepareContent()
	sendNotification()
}

// BaseOTP is the base struct that implements the common methods
type BaseOTP struct{}

// generateOTP generates a random OTP
func (b *BaseOTP) generateOTP(length int) string {
	rand.Seed(time.Now().UnixNano())
	otp := ""
	for i := 0; i < length; i++ {
		otp += fmt.Sprintf("%d", rand.Intn(10))
	}
	fmt.Println("Generated OTP:", otp)
	return otp
}

// saveOTP simulates saving the OTP to cache
func (b *BaseOTP) saveOTP(otp string) {
	fmt.Println("OTP", otp, "saved in cache")
}

// SMSOTP is a concrete class that implements OTP
type SMSOTP struct {
	BaseOTP
}

func (s *SMSOTP) Process() {
	otp := s.generateOTP(6)
	s.saveOTP(otp)
	s.prepareContent()
	s.sendNotification()
}

func (s *SMSOTP) prepareContent() {
	fmt.Println("Preparing SMS content")
}

func (s *SMSOTP) sendNotification() {
	fmt.Println("Sending SMS notification")
}

// EmailOTP is a concrete class that implements OTP
type EmailOTP struct {
	BaseOTP
}

func (e *EmailOTP) Process() {
	otp := e.generateOTP(6)
	e.saveOTP(otp)
	e.prepareContent()
	e.sendNotification()
}

func (e *EmailOTP) prepareContent() {
	fmt.Println("Preparing Email content")
}

func (e *EmailOTP) sendNotification() {
	fmt.Println("Sending Email notification")
}

// AuthenticatorOTP is a concrete class that implements OTP
type AuthenticatorOTP struct {
	BaseOTP
}

func (a *AuthenticatorOTP) Process() {
	otp := a.generateOTP(6)
	a.saveOTP(otp)
	a.prepareContent()
	a.sendNotification()
}

func (a *AuthenticatorOTP) prepareContent() {
	fmt.Println("Preparing Authenticator App content")
}

func (a *AuthenticatorOTP) sendNotification() {
	fmt.Println("Sending Authenticator App notification")
}

File: code/design-patterns/behavioral/template-method/python/template_method.py
from abc import ABC, abstractmethod
import random


class OTP(ABC):
    def process(self):
        otp = self.generate_otp()
        self.save_otp(otp)
        self.prepare_content()
        self.send_notification()

    def generate_otp(self, length=6):
        otp = "".join([str(random.randint(0, 9)) for _ in range(length)])
        print(f"Generated OTP: {otp}")
        return otp

    def save_otp(self, otp):
        # Simulate saving OTP in cache
        print(f"OTP {otp} saved in cache")

    @abstractmethod
    def prepare_content(self):
        pass

    @abstractmethod
    def send_notification(self):
        pass


class SMSOTP(OTP):
    def prepare_content(self):
        print("Preparing SMS content")

    def send_notification(self):
        print("Sending SMS notification")


class EmailOTP(OTP):
    def prepare_content(self):
        print("Preparing Email content")

    def send_notification(self):
        print("Sending Email notification")


class AuthenticatorOTP(OTP):
    def prepare_content(self):
        print("Preparing Authenticator App content")

    def send_notification(self):
        print("Sending Authenticator App notification")
